src/main/scala/io/iohk/ethereum/vm/OpCode.scala

```diff
@@ -446,17 +446,22 @@ case object SSTORE extends OpCode(0x55, 2, 0, G_zero) {
 case object JUMP extends OpCode(0x56, 1, 0, G_mid) with ConstGas {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
     val (pos, stack1) = state.stack.pop
-    //TODO: JUMPDEST validation
-    state.withStack(stack1).goto(pos.intValue)
+    if(state.program.validJumpDestinations.contains(pos.intValue))
+      state.withStack(stack1).goto(pos.intValue)
+    else
+      state.withError(InvalidJump)
   }
 }
 
 case object JUMPI extends OpCode(0x57, 2, 0, G_high) with ConstGas {
   protected def exec[W <: WorldStateProxy[W, S], S <: Storage[S]](state: ProgramState[W, S]): ProgramState[W, S] = {
     val (Seq(pos, cond), stack1) = state.stack.pop(2)
     val nextPos = if (!cond.isZero) pos.intValue else state.pc + 1
-    //TODO: JUMPDEST validation
-    state.withStack(stack1).goto(nextPos)
+    val isValidJump = !cond.isZero && state.program.validJumpDestinations.contains(nextPos)
+    if(isValidJump || cond.isZero)
+      state.withStack(stack1).goto(nextPos)
+    else
+      state.withError(InvalidJump)
```

Comment:
You are right, a JUMPDEST opcode is not necessary for when the jump is not done (so we step to the next instruction). Although using `isValidJump` worked in the same way, I agree that it was too confusing and I'll replace it with your suggestion